content/blog/2019-10-27-tbjit-type-specialization/index.md

```diff
@@ -0,0 +1,189 @@
+
+The diagram below presents two traces. The vertical trace (the root trace) is called directly from the interpreter, while the slanted trace is called from the root trace when a certain branch condition is met. These arrangements form tree-like structures called *trace trees*.
+
+<img src="tree-with-branch.png" alt="Interpreter Performance" width="50%">
+
+Jumping to another trace instead of aborting back to the interpreter is much more efficient.
+
+#### Nested Traces
+
+Traces always consist of a single forward path and end on a backwards path. In the case of a loop nest, instructions will be recorded from a single path through both loops. If there are any conditionals that are post-dominated by the outer loop, then the outer loop instructions can be compiled multiple times (one for each full path). This increases the amount of storage required for reach trace.
+
+The authors propose to effectively perform function outlining on nested loops. One trace can effectively call another trace as the interpreter would.
+
+<img src="nest-trees.png" alt="Interpreter Performance" width="50%">
+
+#### Blacklisting
+
+Specific traces are not worth generating and are prevented from being recorded or run.
+
+## Evaluation
+
+The authors evaluate on a MacBook Pro! JavaScript is somewhat of a consumer language rather than a high-performance language, so a consumer-grade MacBook is somewhat relevant. Keeping in the spirit of consumerism, the authors evaluate on a consumer benchmark suite *SunSpider*. These benchmarks are all extremely small (< 250ms), but webpages generally load within this time.
+
+The authors estimate that a bytecode instruction is 4x faster when compiled to machine code (could be multiple machine instructions, which is why this isn't higher). Most benchmarks spend their time natively executing machine code rather than interpreter bytecode. Thus, every compatible benchmark achieved at least some speedup in TraceMonkey over *SpiderMonkey*, the interpreter-only version of TraceMonkey. TraceMonkey lacks support for certain JavaScript primitives and thus could not gain performance on benchmarks containing those primitives.
+
+I would have liked to have seen the performance impact of the many optimizations they described in their paper, particularly the benefit of creating trace-trees and nested trees.
+
+However, the authors report significant overhead in some benchmarks due to the JIT state machine mainly the recording and compilation procedures. They estimate that a native trace must be executed 270 times to justify the overhead. This may seem small, but generally the total run-time of a JavaScript program is also quite small.
+
+## Discussion Questions
+
+Is the short compilation time potentially limiting the performance that could be achieved by traces? Bytecode only 4x when executed natively? What could be done?
+
+Could an ISA be designed that does not require machine instructions to include type information?
```

Comment:
Yeah, there's research on adding type-checking support to the ISA! For example: https://sampa.cs.washington.edu/new/papers/anderson-hpca2011.pdf